### PR TITLE
General development review

### PR DESCRIPTION
**Pull request by Gaechka777**
PR for reviewing features.

### CODE REVIEW

File: data_preparation/data_reader_upd.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+from data_preparation.dataset_preparation_upd import OrderDataset
+from torch.utils.data import Dataset
+from sklearn.preprocessing import MinMaxScaler
+import numpy as np
+from tqdm import tqdm
+import torch
+from torch.nn.utils.rnn import pad_sequence
+from torch.nn.functional import one_hot
+
+class OrderReader(Dataset):
```
Comment (Gaechka777) [inline data_preparation/data_reader_upd.py:10]: Нужны Docstrings для класса: описание, функционал, атрибуты (self параметры).

File: data_preparation/data_reader_upd.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    def __init__(self, data_folder, train_file, test_file, valid_file, look_back, phase):
```
Comment (Gaechka777) [inline data_preparation/data_reader_upd.py:11]:
```suggestion
    def __init__(self, data_folder: str, train_file: str = None, test_file: str = None, valid_file: str = None, look_back: int = 1, phase: str) -> None:
    """Initialise order reader.
    
    Parameters
    ------------
    data_folder: str
        Folder with input data.
    train_file: str, optional (default=None)
        Name of train file.
    ...
    """
```

Comment (Gaechka777) [reply data_preparation/data_reader_upd.py:11]: Это как идея оформления докстрингов, аналогично можно сделать по всему репозиторию.

File: data_preparation/data_reader_upd.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        super(OrderReader, self).__init__()
+        self.order_dataset = OrderDataset(data_folder, train_file, test_file, valid_file, look_back)
+        self.phase = phase
+
+        [train_final, test_final, valid_final], mms, cat_vocab_size, id_vocab_size = \
+            self.order_dataset.preprocess_dataframe()
+        self.mms = mms
+        if self.phase == 'train':
+            self.df_final, self.ind_combinations = train_final, self.order_dataset.window_combinations(train_final)
+        if self.phase == 'test':
```
Comment (Gaechka777) [inline data_preparation/data_reader_upd.py:21]:
```suggestion
        elif self.phase == 'test':
```

File: data_preparation/data_reader_upd.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            self.df_final, self.ind_combinations = test_final, self.order_dataset.window_combinations(test_final)
+        if self.phase == 'valid':
```
Comment (Gaechka777) [inline data_preparation/data_reader_upd.py:23]:
```suggestion
        elif self.phase == 'valid':
```

File: data_preparation/data_reader_upd.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            self.df_final, self.ind_combinations = valid_final, self.order_dataset.window_combinations(valid_final)
+
```
Comment (Gaechka777) [inline data_preparation/data_reader_upd.py:25]:
```suggestion
else:
    raise ValueError(f"Incorrect phase, expected one of train/valid/test, get {self.phase}")
```

File: data_preparation/data_reader_upd.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        # num_arr consists of summed amount and dt
+        self.cat_arr, self.mask_cat, \
+        self.current_cat, self.mask_current_cat, self.onehot_current_cat, \
+        self.num_arr, self.id_arr, self.target = self.arrange_matrix()
+
+    def arrange_matrix(self):
+        # full_data_len x look_back x max_day_len0
+        cat_arr = pad_sequence([pad_sequence([torch.tensor(self.df_final.loc[ref_i, self.order_dataset.categorical], dtype=torch.int64).unsqueeze(1)
+                                for ref_i in ref_points], padding_value=self.cat_padding_value).squeeze()
+                  for ref_points, pred_point in self.ind_combinations], padding_value=self.cat_padding_value).transpose(1, 0).transpose(2, 1)
```
Comment (Gaechka777) [inline data_preparation/data_reader_upd.py:41]:
```suggestion
                  for ref_points, pred_point in self.ind_combinations], padding_value=self.cat_padding_value).transpose(1, 2, 0)
```

File: data_preparation/data_reader_upd.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+                      for ref_points, pred_point in self.ind_combinations], padding_value=self.cat_padding_value).squeeze().transpose(1, 0)
+
+        # full_data_len x max_day_len1 x 1
+        mask_current_cat = torch.tensor(~(current_cat == self.cat_padding_value), dtype=torch.int64).unsqueeze(2)
+
+        # full_data_len x cat_vocab_size+1
+        onehot_current_cat = torch.sum(one_hot(current_cat, num_classes=self.cat_vocab_size+1) * mask_current_cat, dim=1)
+
+        # full_data_len x look_back x 2
+        num_arr = torch.stack((torch.tensor([[np.sum(self.df_final.loc[ref_i, self.order_dataset.amount]) for ref_i in ref_points]
```
Comment (Gaechka777) [inline data_preparation/data_reader_upd.py:57]: TODO подумать как сделать лучше, чем сумма

File: data_preparation/data_reader_upd.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                                             for ref_points, pred_point in self.ind_combinations], dtype=torch.float32),
+                               torch.tensor([self.df_final.loc[ref_points, self.order_dataset.dt].values.tolist()
+                                            for ref_points, pred_point in self.ind_combinations], dtype=torch.float32)), dim=2)
+
+        # full_data_len x 1
+        id_arr = torch.tensor([self.df_final.loc[ref_points[0], self.order_dataset.id]
+                              for ref_points, pred_point in self.ind_combinations], dtype=torch.int64).reshape(-1, 1)
+
+        # full_data_len x max_day_len1
```
Comment (Gaechka777) [inline data_preparation/data_reader_upd.py:66]: delete useless comments

File: data_preparation/data_reader_upd.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                 for ref_points, pred_point in self.ind_combinations], padding_value=self.amount_padding_value).squeeze().transpose(1, 0)
+
+
+        return cat_arr, mask_cat, current_cat, mask_current_cat, onehot_current_cat, num_arr, id_arr, target
+
+    def __len__(self):
+        return len(self.ind_combinations)
+
+    def __getitem__(self, index):
+        return [self.cat_arr[index, :, :], self.mask_cat[index, :, :],
```
Comment (Gaechka777) [inline data_preparation/data_reader_upd.py:89]: А зачем эти ":" ? Просто индекса же достаточно

File: data_preparation/dataset_preparation_upd.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+
+class OrderDataset:
+    def __init__(self, data_folder, train_file, test_file, valid_file, look_back):
+        self.data_folder = data_folder
+        self.train_file = train_file
+        self.test_file = test_file
+        self.valid_file = valid_file
+        self.look_back = look_back
+
+        # # final columns: id, categorical, amount, date, dt
```
Comment (Gaechka777) [inline data_preparation/dataset_preparation_upd.py:19]: useless comms

File: data_preparation/dataset_preparation_upd.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+
+        mms_dt = get_fitted_scaler(*processed_datasets, self.dt)
+        for processed_df in processed_datasets:
+            processed_df[self.dt] = mms_dt.transform(processed_df[self.dt].values.reshape(-1, 1))
+
+        if self.target == self.amount:
+            return processed_datasets, mms_amount, cat_vocab_size, id_vocab_size
+        if self.target == self.dt:
+            return processed_datasets, mms_dt, cat_vocab_size, id_vocab_size
+
```
Comment (Gaechka777) [inline data_preparation/dataset_preparation_upd.py:125]:
```suggestion
else
    raise NotImplementedError
```

File: data_preparation/dataset_preparation_upd.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+                      for ref_points, pred_point in test_ind_combinations for j in range(len(test_final.loc[pred_point, self.categorical]))])
+        X_test = np.concatenate((ohe.transform(X_test_arr[:, :2]), X_test_arr[:, 2].reshape(-1, 1)), axis=1)
+        y_test = np.array([[test_final.loc[pred_point, self.amount][j]]
+                            for ref_points, pred_point in test_ind_combinations for j in range(len(test_final.loc[pred_point, self.categorical]))])
+
+        return [X_train, y_train, X_test, y_test], mms
+
+    def construct_features_joining(self):
+        """
+        Construct features and target for Machine Learning algorithms.
```
Comment (Gaechka777) [inline data_preparation/dataset_preparation_upd.py:169]: А в чем отличие от обычного construct_features? (Докстринги одинаковые)

File: data_preparation/dataset_preparation_upd.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+                ohe_cat.fit_transform(np.array(df_final.loc[pred_point, self.categorical][j]).reshape(1, -1)).reshape(-1).tolist()
+                for ref_points, pred_point in tqdm(ind_combinations) for j in range(len(df_final.loc[pred_point, self.categorical]))])
+            y = np.array([[df_final.loc[pred_point, self.amount][j]]
+                          for ref_points, pred_point in ind_combinations for j in range(len(df_final.loc[pred_point, self.categorical]))])
+            ml_datasets.append(X)
+            ml_datasets.append(y)
+        return ml_datasets, mms
+
+
+    # def construct_features(self):
```
Comment (Gaechka777) [inline data_preparation/dataset_preparation_upd.py:197]: del pls

Comment (Gaechka777) [reply data_preparation/dataset_preparation_upd.py:197]: Ну или в отдельную веточку, если в будущем пригодятся

File: data_preparation/data_reader_simultaneous.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
+from data_preparation.dataset_preparation_upd import OrderDataset
+from torch.utils.data import Dataset
+from sklearn.preprocessing import MinMaxScaler
+import numpy as np
+from tqdm import tqdm
+import torch
+from torch.nn.utils.rnn import pad_sequence
+from torch.nn.functional import one_hot
+
+class OrderReader(Dataset):
```
Comment (Gaechka777) [inline data_preparation/data_reader_simultaneous.py:10]:
```suggestion
from data_reader import OrderReader

class OrderReader_simultaneous(OrderReader):
```

Comment (Gaechka777) [reply data_preparation/data_reader_simultaneous.py:10]: Позволит тебе не копипастить код и сделать классовую иерархию

File: models/regression/lstm_material.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        super(LSTMEncoding, self).__init__()
+
+        self.lstm_input_dim = lstm_input_dim
+        self.lstm_hidden_dim = lstm_hidden_dim
+
+        self.lstm = nn.LSTM(input_size=lstm_input_dim, hidden_size=lstm_hidden_dim, batch_first=True)
+
+    def forward(self, x):
+        x_out, (_, _) = self.lstm(x)
+        x_last_state = x_out[:, -1, :]
```
Comment (Gaechka777) [inline models/regression/lstm_material.py:16]:
```suggestion
        x_last_state = x_out[:, -1]
```

File: models/regression/lstm_material.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+        x_encoding_cat = self.lstm_encoding_cat(x_cat_emb_sum)  # [batch_size, lstm_hidden_dim]
+
+        x_num1 = self.linear_num_feat(num_arr)  # [batch_size, look_back, linear_num_feat_dim]
+        x_encoding_num = self.lstm_encoding_num(x_num1)  # [batch_size, lstm_hidden_dim]
+
+        x_id_emb = self.id_embedding(id_arr).squeeze(1)  # [batch_size, id_embedding_dim]
+        x_concat = torch.cat((x_encoding_cat, x_encoding_num, x_id_emb),
+                             dim=1)  # [batch_size, 2*lstm_hidden_dim+id_embedding_dim]
+
```
Comment (Gaechka777) [inline models/regression/lstm_material.py:56]: Скорее всего BatchNorm и/или DropOut после каждого линейного слоя улучшит качество.
Также можно подумать в сторону Conv1D вместо линейных слоёв.

File: models/regression/lstm_material.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
+import torch
+from torch import nn
+
+
+class LSTMEncoding(nn.Module):
```
Comment (Gaechka777) [inline models/regression/lstm_material.py:5]: Кажется можно было просто from .lstm_model import LSTMEncoding

File: models/regression/lstm_multidim_amount_upd.py [hunk 18]
```diff
@@ -0,0 +1,70 @@
+import torch
+from torch import nn
+
+
+class LSTMEncoding(nn.Module):
```
Comment (Gaechka777) [inline models/regression/lstm_multidim_amount_upd.py:5]: from .lstm_model import this

File: launch_scripts/train_lstm.py [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+    const_embedding_dim = 256
+    linear_embedding1_dim = 128
+    linear_embedding2_dim = 64
+    linear_concat1 = 32
+
+    net = RegressionNet(lstm_input_dim, lstm_hidden_dim, linear_encoding1_dim, linear_encoding2_dim,
+                        vocab_size0, vocab_size1, vocab_size2, const_embedding_dim,
+                        linear_embedding1_dim, linear_embedding2_dim,
+                        linear_concat1).to(device)
+    optimizer = torch.optim.Adam(net.parameters(), lr=1e-4)
```
Comment (Gaechka777) [inline launch_scripts/train_lstm.py:45]: можно попробовать AdamW

File: launch_scripts/train_lstm.py [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+    # scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode='min', factor=0.5, patience=3)
```
Comment (Gaechka777) [inline launch_scripts/train_lstm.py:46]: вообще шедулер нужен, позволяет лучше сойтись нейронке

File: launch_scripts/train_lstm.py [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+            batch_const_feat.to(device), batch_changing_feat.to(device), batch_target.to(device)
+            optimizer.zero_grad()
+            predicted_amounts = net(batch_changing_feat.float(), batch_const_feat.int())
+            loss = regr_loss(predicted_amounts, batch_target.float())
+            epoch_train_loss += loss.item()
+            loss.backward()
+            optimizer.step()
+
+        print(f'Epoch {epoch}/{num_epochs} || Train loss {epoch_train_loss}')
+        net.train(False)
```
Comment (Gaechka777) [inline launch_scripts/train_lstm.py:72]:
```suggestion
        net.eval()
```

File: launch_scripts/train_lstm.py [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+        test_predicted_amounts = []
+        test_gt_amounts = []
+
+        for batch_ind, (batch_const_feat, batch_changing_feat, batch_target) in enumerate(test_dataloader):
+            batch_const_feat, batch_changing_feat, batch_target = \
+            batch_const_feat.to(device), batch_changing_feat.to(device), batch_target.to(device)
+            test_predicted_amounts.extend(net(batch_changing_feat.float(), batch_const_feat.int()).squeeze().detach().cpu().numpy().tolist())
+            test_gt_amounts.extend(batch_target.squeeze().detach().cpu().numpy().tolist())
+
+        r2_metric_test = r2_score(test_dataset.mms_amount.inverse_transform(np.array(test_gt_amounts, dtype=np.float64).reshape(-1, 1)),
```
Comment (Gaechka777) [inline launch_scripts/train_lstm.py:82]: Попробовать без объединения в один большой вектор

File: launch_scripts/train_lstm_simultaneous.py [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+             batch_num_arr, batch_id_arr, batch_target_amount, batch_target_cat] = batch_arrays
+            output_amount, output_material = net(batch_cat_arr, batch_mask_cat, batch_num_arr, batch_id_arr)
+
+            nonzero_indices = batch_onehot_current_cat.reshape(-1).nonzero()
+            batch_predicted_amounts = (output_amount * batch_onehot_current_cat).reshape(-1)[nonzero_indices]
+            batch_gt_amounts = batch_target_amount.reshape(-1)[(batch_target_amount.reshape(-1) - train_dataset.amount_padding_value).nonzero()]
+
+            # loss = regr_loss(batch_predicted_amounts, batch_gt_amounts) + \
+            #        alpha * multilabel_crossentropy_loss(output_material, batch_onehot_current_cat)
+            loss = 10**6 * regr_loss(batch_predicted_amounts, batch_gt_amounts) + \
```
Comment (Gaechka777) [inline launch_scripts/train_lstm_simultaneous.py:192]: почему 10^6?)))
